src/feature/dirclient/dirclient.c

```diff
@@ -759,6 +760,10 @@ connection_dir_client_request_failed(dir_connection_t *conn)
     log_info(LD_DIR, "Giving up on downloading microdescriptors from "
              "directory server at '%s'; will retry", conn->base_.address);
     connection_dir_download_routerdesc_failed(conn);
+  } else if (TO_CONN(conn)->purpose == DIR_PURPOSE_UPLOAD_HSDESC &&
+             conn->hs_ident) {
```

Comment:
Has this feature been tested and if yes, how? The lack of unittests is a bit distressing.

Reply:
Testing that specific feature is sorta hard in a regression test environment (let say Chutney). One would need to close the channel link in the HSDir upload circuit right when the descriptor uploads...

Even with unit test, this is difficult to test. Uploading a descriptor is async (request <-> reply) so at best the unit test would test the correctness of the function based on what we pass it. (which most of the code is already tested except, I give you that, the new `hs_service_reupload_desc_to_dir()`).

I'll add a unit test that actually confirms that if we end up in the code path of "connection ended badly", we end up re-uploading.

But I won't be able to test that up to the directory upload :S.

Reply:
Suggestion of real life test: Can you test that by setting up an onion service with a small descriptor upload time, and then shutting down the network before it needs to upload descriptor?